File: backend/user-service/app/services/user_service.py
"""User service for business logic."""

from datetime import datetime

from bookmyshow_common.exceptions import (
    ResourceAlreadyExistsException,
    ResourceNotFoundException,
)
from bookmyshow_common.security import PasswordHasher

from app.models import User
from app.repositories.user_repository import UserRepository
from app.schemas.user import RegisterRequest, UpdateProfileRequest, UserProfileResponse


class UserService:
    """Service for user-related business logic."""

    def __init__(self, repository: UserRepository):
        """Initialize user service.
        
        Args:
            repository: User repository instance
        """
        self.repository = repository
        self.password_hasher = PasswordHasher()

    async def register_user(self, request: RegisterRequest) -> UserProfileResponse:
        """Register a new user.
        
        Args:
            request: User registration request
            
        Returns:
            User profile response
            
        Raises:
            ResourceAlreadyExistsException: If user with email already exists
        """
        # Check if user already exists
        if await self.repository.exists_by_email(request.email):
            raise ResourceAlreadyExistsException(
                f"User with email {request.email} already exists"
            )

        # Create user
        user = User(
            email=request.email,
            password_hash=self.password_hasher.hash_password(request.password),
            first_name=request.first_name,
            last_name=request.last_name,
            phone_number=request.phone_number,
            role="CUSTOMER",
            is_active=True,
            is_email_verified=False,
        )

        saved_user = await self.repository.create_user(user)
        return self._map_to_user_profile_response(saved_user)

    async def get_user_profile(self, user_id: int) -> UserProfileResponse:
        """Get user profile by ID.
        
        Args:
            user_id: User ID
            
        Returns:
            User profile response
            
        Raises:
            ResourceNotFoundException: If user not found
        """
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise ResourceNotFoundException(message=f"User not found with id: {user_id}")
        return self._map_to_user_profile_response(user)

    async def get_user_by_email(self, email: str) -> User:
        """Get user by email.
        
        Args:
            email: User email
            
        Returns:
            User model
            
        Raises:
            ResourceNotFoundException: If user not found
        """
        user = await self.repository.get_user_by_email(email)
        if not user:
            raise ResourceNotFoundException(message=f"User not found with email: {email}")
        return user

    async def get_user_by_id(self, user_id: int) -> User:
        """Get user by ID.
        
        Args:
            user_id: User ID
            
        Returns:
            User model
            
        Raises:
            ResourceNotFoundException: If user not found
        """
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise ResourceNotFoundException(message=f"User not found with id: {user_id}")
        return user

    async def update_user_profile(
        self, user_id: int, request: UpdateProfileRequest
    ) -> UserProfileResponse:
        """Update user profile.
        
        Args:
            user_id: User ID
            request: Update profile request
            
        Returns:
            Updated user profile response
            
        Raises:
            ResourceNotFoundException: If user not found
        """
        user = await self.get_user_by_id(user_id)

        # Update fields if provided
        if request.first_name is not None:
            user.first_name = request.first_name
        if request.last_name is not None:
            user.last_name = request.last_name
        if request.phone_number is not None:
            user.phone_number = request.phone_number

        updated_user = await self.repository.update_user(user)
        return self._map_to_user_profile_response(updated_user)

    async def update_last_login(self, user_id: int) -> None:
        """Update user's last login timestamp.
        
        Args:
            user_id: User ID
        """
        await self.repository.update_last_login(user_id, datetime.utcnow())

    def _map_to_user_profile_response(self, user: User) -> UserProfileResponse:
        """Map User model to UserProfileResponse.
        
        Args:
            user: User model instance
            
        Returns:
            User profile response
        """
        return UserProfileResponse(
            id=user.id,
            email=user.email,
            first_name=user.first_name,
            last_name=user.last_name,
            phone_number=user.phone_number,
            role=user.role,
            is_active=user.is_active,
            is_email_verified=user.is_email_verified,
            created_at=user.created_at,
            last_login_at=user.last_login_at,
        )

File: backend/user-service/src/main/java/com/bookmyshow/user/service/UserService.java
package com.bookmyshow.user.service;

import com.bookmyshow.common.exception.ResourceAlreadyExistsException;
import com.bookmyshow.common.exception.ResourceNotFoundException;
import com.bookmyshow.user.dto.RegisterRequest;
import com.bookmyshow.user.dto.UpdateProfileRequest;
import com.bookmyshow.user.dto.UserProfileResponse;
import com.bookmyshow.user.entity.User;
import com.bookmyshow.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Service
@Transactional
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserProfileResponse registerUser(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new ResourceAlreadyExistsException(
                "User with email " + request.getEmail() + " already exists"
            );
        }

        User user = new User();
        user.setEmail(request.getEmail());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setFirstName(request.getFirstName());
        user.setLastName(request.getLastName());
        user.setPhoneNumber(request.getPhoneNumber());
        user.setRole("CUSTOMER");
        user.setIsActive(true);
        user.setIsEmailVerified(false);

        User savedUser = userRepository.save(user);
        return mapToUserProfileResponse(savedUser);
    }

    public UserProfileResponse getUserProfile(Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));
        return mapToUserProfileResponse(user);
    }

    public User getUserByEmail(String email) {
        return userRepository.findByEmail(email)
            .orElseThrow(() -> new ResourceNotFoundException("User not found with email: " + email));
    }

    public User getUserById(Long userId) {
        return userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));
    }

    public UserProfileResponse updateUserProfile(Long userId, UpdateProfileRequest request) {
        User user = getUserById(userId);

        if (request.getFirstName() != null) {
            user.setFirstName(request.getFirstName());
        }
        if (request.getLastName() != null) {
            user.setLastName(request.getLastName());
        }
        if (request.getPhoneNumber() != null) {
            user.setPhoneNumber(request.getPhoneNumber());
        }

        User updatedUser = userRepository.save(user);
        return mapToUserProfileResponse(updatedUser);
    }

    public void updateLastLogin(Long userId) {
        User user = getUserById(userId);
        user.setLastLoginAt(LocalDateTime.now());
        userRepository.save(user);
    }

    private UserProfileResponse mapToUserProfileResponse(User user) {
        return new UserProfileResponse(
            user.getId(),
            user.getEmail(),
            user.getFirstName(),
            user.getLastName(),
            user.getPhoneNumber(),
            user.getRole(),
            user.getIsActive(),
            user.getIsEmailVerified(),
            user.getCreatedAt(),
            user.getLastLoginAt()
        );
    }
}
